### PR TITLE
Fix #12

### LINKED ISSUE #12: Issue with ssa's handling of reassignment of function argument
Test available on the branch `func-arg-issue`, here's the test code
```python
def test_reassign_arg():
    def bar(x):
        return x

    @end_rewrite()
    @ssa()
    @begin_rewrite()
    def foo(a, b):
        if b:
            a = bar(a)
        return a
    assert inspect.getsource(foo) == """\
def foo(a, b):
    a0 = bar(a)
    a1 = a0 if b else a
    __return_value0 = a1
    return __return_value0
"""

```

Fails with

```
E       AssertionError: assert 'def foo(a, b...turn_value0\n' == 'def foo(a, b)...turn_value0\n'
E           def foo(a, b):
E         -     a0 = bar(a0)
E         ?               -
E         +     a0 = bar(a)
E               a1 = a0 if b else a
E               __return_value0 = a1
E               return __return_value0
```

It looks like the first assignment to the name a0, which is one of the function arguments, is also updating the first read of the name on the RHS of the assignment expression, instead it should be `a` (the original parameter value).

### PR DESCRIPTION
**Pull request by leonardt**

### CODE REVIEW

**Comment by leonardt** [comment]
This actually introduces a different bug when looking at nested function calls

```
def test_double_nested_function_call():
    def bar(x):
        return x

    def baz(x):
        return x + 1

    @end_rewrite()
    @ssa()
    @begin_rewrite()
    def foo(a, b, c):
        if b:
            a = bar(a)
        else:
            a = bar(a)
        if c:
            b = bar(b)
        else:
            b = bar(b)
        return a, b
    print(inspect.getsource(foo))
    assert inspect.getsource(foo) == """\
def foo(a, b, c):
    a0 = bar(a)
    a1 = bar(a)
    a2 = a0 if b else a1
    b0 = bar(b)
    b1 = bar(b)
    b2 = b0 if c else b1
    __return_value0 = a2
    return __return_value0
"""
```

fails with

```
______________________________________ test_double_nested_function_call _______________________________________

    def test_double_nested_function_call():
        def bar(x):
            return x

        def baz(x):
            return x + 1

        @end_rewrite()
        @ssa()
        @begin_rewrite()
        def foo(a, b, c):
            if b:
                a = bar(a)
            else:
                a = bar(a)
            if c:
                b = bar(b)
            else:
                b = bar(b)
            return a, b
        print(inspect.getsource(foo))
>       assert inspect.getsource(foo) == """\
    def foo(a, b, c):
        a0 = bar(a)
        a1 = bar(a)
        a2 = a0 if b else a1
        b0 = bar(b)
        b1 = bar(b)
        b2 = b0 if c else b1
        __return_value0 = a2
        return __return_value0
    """
E       AssertionError: assert 'def foo(a, b...turn_value0\n' == 'def foo(a, b,...turn_value0\n'
E         Skipping 70 identical leading characters in diff, use -v to show
E           e a1
E         -     bar0 = bar if b else bar
E         -     b0 = bar0(b)
E         ?             -
E         +     b0 = bar(b)
E         -     b1 = bar0(b)...
E
E         ...Full output truncated (8 lines hidden), use '-vv' to show

tests/test_ssa.py:166: AssertionError
-------------------------------------------- Captured stdout call ---------------------------------------------
def foo(a, b, c):
    a0 = bar(a)
    a1 = bar(a)
    a2 = a0 if b else a1
    bar0 = bar if b else bar
    b0 = bar0(b)
    b1 = bar0(b)
    b2 = b0 if c else b1
    __return_value0 = a2, b2
    return __return_value0
```

**Comment by leonardt** [comment]
It seems to be ssa-ifying the function name, so perhaps there's some logic that relies on the fact that assignments are visited in the other order.

**Comment by leonardt** [comment]
(reverting the commit to swap the assign ordering fixes the issue with the function name being ssa-ified)

File: ast_tools/passes/ssa.py [hunk 1]
```diff
@@ -62,6 +62,11 @@ def visit(self, node: ast.AST) -> ast.AST:
         else:
             return super().visit(node)
 
+    def visit_Assign(self, node):
```
Comment (cdonovick) [inline ast_tools/passes/ssa.py]: Im confused how is this not just `generic_visit`?

Comment (cdonovick) [reply ast_tools/passes/ssa.py]: I should have read the issue first.

**Comment by cdonovick** [comment]
Visiting the RHS is important.

if visiting LHS first:
```
def foo()
   x = 1
   x = x + 1
```
becomes
```
def foo()
   x0 = 1
   x1 = x1 + 1
```

I'm confused as why function names would  be ssa-ified visiting LHS first.  Only names in the store context should be ssa-ified. I look into it.

**Pull request merged**